.travis.yml

```diff
@@ -1,100 +1,75 @@
- - sed -i -e '1,/^Resolving /d' installplan.txt; cat installplan.txt
 
-# check whether current requested install-plan matches cached package-db snapshot
- - if diff -u $HOME/.cabsnap/installplan.txt installplan.txt;
-   then
-     echo "cabal build-cache HIT";
-     rm -rfv .ghc;
-     cp -a $HOME/.cabsnap/ghc $HOME/.ghc;
-     cp -a $HOME/.cabsnap/lib $HOME/.cabsnap/share $HOME/.cabsnap/bin $HOME/.cabal/;
-   else
-     echo "cabal build-cache MISS";
-     rm -rf $HOME/.cabsnap;
-     mkdir -p $HOME/.ghc $HOME/.cabal/lib $HOME/.cabal/share $HOME/.cabal/bin;
-     cabal install --only-dependencies --enable-tests --enable-benchmarks;
-   fi
+install:
+  - echo "$(ghc --version) [$(ghc --print-project-git-commit-id 2> /dev/null || echo '?')]"
+  - if [ -f configure.ac ]; then autoreconf -i; fi
+  - set -ex
+  - stack --no-terminal --install-ghc $ARGS test --bench --only-dependencies
+  - set +ex
 
-# snapshot package-db on cache miss
- - if [ ! -d $HOME/.cabsnap ];
-   then
-      echo "snapshotting package-db to build-cache";
-      mkdir $HOME/.cabsnap;
-      cp -a $HOME/.ghc $HOME/.cabsnap/ghc;
-      cp -a $HOME/.cabal/lib $HOME/.cabal/share $HOME/.cabal/bin installplan.txt $HOME/.cabsnap/;
-   fi
 
-# Here starts the actual work to be performed for the package under test;
-# any command which exits with a non-zero exit code causes the build to fail.
 script:
- - if [ -f configure.ac ]; then autoreconf -i; fi
- - cabal configure --enable-tests --enable-benchmarks -v2  # -v2 provides useful information for debugging
- - cabal build   # this builds all libraries and executables (including tests/benchmarks)
- - cabal test
- - cabal check
- - cabal sdist   # tests that a source-distribution can be generated
-
-# Check that the resulting source distribution can be built & installed.
-# If there are no other `.tar.gz` files in `dist`, this can be even simpler:
-# `cabal install --force-reinstalls dist/*-*.tar.gz`
- - SRC_TGZ=$(cabal info . | awk '{print $2;exit}').tar.gz &&
-   (cd dist && cabal install --force-reinstalls "$SRC_TGZ")
+  - export PATH="${PATH}:$(pwd)/artifacts"
+  - stack build --copy-bins --local-bin-path ./bin;
+  - get_cabal_version() { cat $1/$1.cabal | grep '^Version: ' | gsed -e 's/^Version: //g'; }
+  - mk_release_name() { echo "$1-$(get_cabal_version $1)-x86_64-macos.tar.bz2"; }
+  - tar -jcvf $(mk_release_name dhall) bin/dhall
+  - tar -jcvf $(mk_release_name dhall-json) bin/dhall-to-json bin/dhall-to-yaml bin/json-to-dhall bin/yaml-to-dhall
+  - tar -jcvf $(mk_release_name dhall-bash) bin/dhall-to-bash
+  - tar -jcvf $(mk_release_name dhall-lsp-server) bin/dhall-lsp-server
 
 before_deploy:
-  - tar --create --file "$TRAVIS_OS_NAME.tar" --files-from /dev/null
-  - tar --append --file "$TRAVIS_OS_NAME.tar" --directory dist/build/dhall dhall
-  - gzip "$TRAVIS_OS_NAME.tar"
+  - stack build --copy-bins --local-bin-path ./bin;
+  # TODO: move the tar operations here
```

Comment:
The `before_deploy` phase only runs when the `deploy` phase runs, right? Since we only run the `deploy` phase on tags, it'd be nice to know whether the book keeping of the deploy works separate from the actual process of deploying.

I'd be willing to suggest that we even move this stuff into a phase that [runs all the time][job lifecycle]: `script` or `after_success`. Nothing really needs to change for this PR, though. The way it's laid out currently solves the problem at hand, and that's more valuable than toiling in perfection.

[job lifecycle]: https://docs.travis-ci.com/user/job-lifecycle/

Reply:
Would this mean _not_ moving the tar operations as the TODO suggests?

Reply:
Right.

Reply:
@joneshf do you know if the env is shared between stages (so that we can build in `script` and have a no-op here) or should I duplicate the tar stuff?

Reply:
I think that's what the [`skip_cleanup`][] flag controls. There's only that blurb about it though.

[`skip_cleanup`]: https://docs.travis-ci.com/user/job-lifecycle/#deploying-your-code

Reply:
@joneshf cool thanks, let's try that and fix it in next release if it doesn't work